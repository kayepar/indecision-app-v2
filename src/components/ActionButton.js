import React from 'react';
import { useOptionsContext } from '../context/optionsContext';

const ActionButton = (props) => {
    const { options } = useOptionsContext();

    console.log(options);

    const handlePickOption = () => {
        const randomNumber = Math.floor(Math.random() * options.length);
        const pickedOption = options[randomNumber];

        props.updatePickedOption(pickedOption);
    };

    return (
        <div>
            <button id="big-button" className="big-button" onClick={handlePickOption} disabled={options.length < 1}>
                Choose for me
            </button>
        </div>
    );
};

export default ActionButton;
