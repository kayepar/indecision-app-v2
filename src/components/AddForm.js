import React, { useState } from 'react';
import { useOptionsContext } from '../context/optionsContext';

const AddForm = () => {
    const [error, setError] = useState(undefined);
    const { options, optionsDispatch } = useOptionsContext();

    const handleAddOption = (e) => {
        e.preventDefault();

        const option = e.target.elements.option.value.trim();
        const formError = validateForm(option);

        if (formError) {
            setError(formError);
        } else {
            optionsDispatch({ type: 'ADD_OPTION', option });
            e.target.elements.option.value = '';
            setError(undefined);
        }
    };

    const validateForm = (option) => {
        let error;

        if (!option) {
            error = 'Enter a valid option';
        } else if (ifOptionExists(option)) {
            error = `This option already exists (# ${getOptionIndex(option)})`;
        }

        return error;
    };

    const ifOptionExists = (option) => {
        return options.indexOf(option) > -1;
    };

    const getOptionIndex = (option) => {
        return options.findIndex((item) => item === option) + 1;
    };

    return (
        <div className="add-option-container">
            {error && <p className="add-option-error">{error}</p>}
            <form className="add-option" onSubmit={handleAddOption}>
                <input className="add-option__input" type="text" name="option" autoComplete="off" autoFocus></input>
                <button className="button">Add Option</button>
            </form>
        </div>
    );
};

export default AddForm;
