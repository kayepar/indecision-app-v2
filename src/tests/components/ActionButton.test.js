import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ActionButton from '../../components/ActionButton';
import OptionsProvider from '../../context/optionsContext';
import AddForm from '../../components/AddForm';

afterEach(cleanup);

const updatePickedOptionSpy = jest.fn();

const renderComponents = () => {
    return render(
        <OptionsProvider>
            <ActionButton updatePickedOption={updatePickedOptionSpy} />
            <AddForm />
        </OptionsProvider>
    );
};

const addOption = (option) => {
    userEvent.type(screen.getByRole('textbox'), option);
    userEvent.click(screen.getByRole('button', { name: 'Add Option' }));
};

describe('Tests for ActionButton', () => {
    test('Should correctly render the component', () => {
        const { asFragment } = renderComponents();

        expect(asFragment()).toMatchSnapshot();
    });

    test('If there are no options, should render the button as disabled', () => {
        const { getByRole } = renderComponents();
        const chooseForMeButton = getByRole('button', { name: 'Choose for me' });

        expect(chooseForMeButton).toBeDisabled();
    });

    test('If option is available, should render the button as enabled', () => {
        const { getByRole } = renderComponents();
        const chooseForMeButton = getByRole('button', { name: 'Choose for me' });

        addOption('Javascript');

        expect(chooseForMeButton).toBeEnabled();
    });

    test('If button is clicked, should call updatePickedOption prop function', () => {
        const { getByRole } = renderComponents();
        const chooseForMeButton = getByRole('button', { name: 'Choose for me' });

        addOption('HTML5');
        userEvent.click(chooseForMeButton);

        expect(updatePickedOptionSpy).toBeCalledWith('HTML5');
    });
});
