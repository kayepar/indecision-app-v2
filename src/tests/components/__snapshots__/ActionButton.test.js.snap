// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for ActionButton Should correctly render the component 1`] = `
<DocumentFragment>
  <div>
    <button
      class="big-button"
      disabled=""
      id="big-button"
    >
      Choose for me
    </button>
  </div>
  <div
    class="add-option-container"
  >
    <form
      class="add-option"
    >
      <input
        autocomplete="off"
        class="add-option__input"
        name="option"
        type="text"
      />
      <button
        class="button"
      >
        Add Option
      </button>
    </form>
  </div>
</DocumentFragment>
`;
